src/renderer.ts

```diff
@@ -351,7 +358,16 @@ class Renderer extends EventEmitter<RendererEvents> {
     const { scrollLeft, scrollWidth, clientWidth } = this.scrollContainer
     const len = channelData[0].length
     const scale = len / scrollWidth
-    const viewportWidth = Math.min(Renderer.MAX_CANVAS_WIDTH, clientWidth)
+    let viewportWidth = Math.min(Renderer.MAX_CANVAS_WIDTH, clientWidth)
+
+    if (options.barWidth || options.barGap) {
+      const barWidth = options.barWidth ?? 1
+      const barGap = options.barGap ?? barWidth / 2
+      viewportWidth = this.closestInteger(
+        Math.min(Renderer.MAX_CANVAS_WIDTH, clientWidth) + barWidth / barGap,
+        barWidth + barGap,
+      )
+    }
```

Comment:
How about something like this?

```suggestion
    if (options.barWidth || options.barGap) {
        const pixelRatio = window.devicePixelRatio || 1
        const barWidth = options.barWidth ? options.barWidth * pixelRatio : 1
        const barGap = options.barGap ? options.barGap * pixelRatio : options.barWidth ? barWidth / 2 : 0
        const remainder = viewportWidth % (barWidth + barGap)
        // viewportWidth must fit a whole number of bars and gaps
        if (remainder > 0) {
          viewportWidth += (barWidth + barGap - remainder)
        }
      )
    }
```